### PR TITLE
[narando] add new extractor

### PR DESCRIPTION
**Pull request by ealgase**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This implements an extractor for narando as specified in #18252 .  This also adds an extractor for [narando:player], which can be used in the future for sites which embed the narando player.

### CODE REVIEW

File: youtube_dl/extractor/narando.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            'title': 'An  ihrem  Selbstlob  erkennt  man  sie',
+            'url': 'https://static.narando.com/sounds/10492/original.mp3',
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage('https://narando.com/articles/' + video_id, video_id)
+        title = self._html_search_regex(r'<h1 class="visible-xs h3">(.+?)</h1>', webpage, 'title')
+        player_id = self._html_search_regex(r'[\n\r].*https:\/\/narando.com\/r\/\s*([^"]*)', webpage, 'player_id')
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: `[\n\r].*` does not make much sense. `\s*` same. No escapes for forward slash. Does not make sense to capture empty string.

File: youtube_dl/extractor/narando.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+
+class NarandoPlayerIE(InfoExtractor):
+    IE_NAME = "narando:player"
+    _VALID_URL = r'https://narando.com/widget\?r=(?P<id>\w+)'
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Escape dots.

File: youtube_dl/extractor/narando.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            'id': 'b2t4t789kxgy9g7ms4rwjvvw',
+            'ext': 'mp3',
+            'title': 'An  ihrem  Selbstlob  erkennt  man  sie',
+            'url': 'https://static.narando.com/sounds/10492/original.mp3',
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage('https://narando.com/widget?r=' + video_id, video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Same.

File: youtube_dl/extractor/narando.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        title = self._html_search_regex(r'<title>narando \| (.+?)</title>', webpage, 'title')
+        download_url = self._html_search_regex(r'.<div class="stream_url hide">\s*([^?]*)', webpage, 'download_url')
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Same.

Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Capture between tags.

File: youtube_dl/extractor/narando.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    IE_NAME = "narando:player"
+    _VALID_URL = r'https://narando.com/widget\?r=(?P<id>\w+)'
+    _TEST = {
+        'url': 'https://narando.com/widget?r=b2t4t789kxgy9g7ms4rwjvvw',
+        'md5': 'd20f671f0395bab8f8285d1f6e8f965e',
+        'info_dict': {
+            'id': 'b2t4t789kxgy9g7ms4rwjvvw',
+            'ext': 'mp3',
+            'title': 'An  ihrem  Selbstlob  erkennt  man  sie',
+            'url': 'https://static.narando.com/sounds/10492/original.mp3',
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: No exact URLs here.

File: youtube_dl/extractor/narando.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        video_id = self._match_id(url)
+        webpage = self._download_webpage('https://narando.com/widget?r=' + video_id, video_id)
+        title = self._html_search_regex(r'<title>narando \| (.+?)</title>', webpage, 'title')
+        download_url = self._html_search_regex(r'.<div class="stream_url hide">\s*([^?]*)', webpage, 'download_url')
+
+        return {
+            'id': video_id,
+            'title': title,
+            'url': download_url,
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: vcodec 'none'.

**Comment by ealgase** [comment]
Thanks for reviewing the code and suggesting improvements, I've implemented them.

File: youtube_dl/extractor/extractors.py [hunk 7]
```diff
@@ -1480,3 +1484,5 @@
 from .zdf import ZDFIE, ZDFChannelIE
 from .zingmp3 import ZingMp3IE
 from .zype import ZypeIE
+
+
```
Comment (dstftw) [inline youtube_dl/extractor/extractors.py]: Remove superfluous whitespace.

**Comment by ealgase** [comment]
@dstftw are there any more things I need to fix or can this be merged?

File: youtube_dl/extractor/narando.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+
+class NarandoPlayerIE(InfoExtractor):
+    IE_NAME = "narando:player"
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Single quotes.

File: youtube_dl/extractor/narando.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+
+class NarandoPlayerIE(InfoExtractor):
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Rename to `NarandoIE`.

File: youtube_dl/extractor/narando.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    IE_NAME = "narando:player"
+    _VALID_URL = r'https://narando\.com/widget\?r=(?P<id>\w+)'
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Does not match https://narando.com/widget?foo=bar&r=b2t4t789kxgy9g7ms4rwjvvw.

File: youtube_dl/extractor/narando.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+            'ext': 'mp3',
+            'title': 'An  ihrem  Selbstlob  erkennt  man  sie',
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        title = self._html_search_regex(r'<span class="clip-title">(.+?)</span>', webpage, 'title')
+        download_url = self._html_search_regex(r'.<div class="stream_url hide">\s*([^?]*)', webpage, 'download_url')
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: What's `.` for?

Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Do not capture empty strings.

File: youtube_dl/extractor/narando.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        return {
+            'id': video_id,
+            'title': title,
+            'url': download_url,
+            'vcodec': 'none',
+        }
+
+
+class NarandoIE(InfoExtractor):
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Rename.

File: youtube_dl/extractor/narando.py (hunk 10 above)
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: Does not match https://narando.com/r/b2t4t789kxgy9g7ms4rwjvvw.

File: youtube_dl/extractor/narando.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+            'ext': 'mp3',
+            'title': 'An  ihrem  Selbstlob  erkennt  man  sie',
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        title = self._html_search_regex(r'<h1 class="visible-xs h3">(.+?)</h1>', webpage, 'title')
+        player_id = self._html_search_regex(r'\s*https://narando.com/r/([^"]*)', webpage, 'player_id')
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: `\s*` is useless here. Again do not capture empty strings.

File: youtube_dl/extractor/narando.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        player_url = 'https://narando.com/widget?r=' + player_id
+
+        return {
```
Comment (dstftw) [inline youtube_dl/extractor/narando.py]: `url_result`.

**Comment by ealgase** [comment]
Thanks for reviewing again.  I've made the requested changes.

**Comment by ealgase** [comment]
@dstftw Can you please take a look at this again?  I've also added thumbnail extraction.

**Pull request closed without merging**